### ISSUE TITLE
JIT with MLIR

### ISSUE THREAD

**Issue by danzimm** [fish-shell/fish-shell#10978]
Hey folks, this issue is basically https://github.com/fish-shell/fish-shell/issues/4316 again, but it might be more feasible now that `fish` is written in `rust`.

I started to spec out implementing a JIT with MLIR with [melior](https://github.com/mlir-rs/melior). I was thinking we could implement this for functions, to start out, and expand to scripts as we see fit. Here's the plan of attack I quickly came up with:

- Pin LLVM somehow (either submodule or a hash like what [triton does](https://github.com/triton-lang/triton/blob/main/cmake/llvm-hash.txt))
- Depend on `melior`
- Implement AST > MLIR lowering
- Implement a JIT cache which `exec.rs` would call out

In order to make implementing this feasible I was thinking we could iteratively add support for each AST node, instead of adding support all at once.

Additionally, to start, I was thinking we could have an explicit flag on a function definition `--jit` to indicate the function should be `jit`d.

As the JIT infra grows we could implement various profiling techniques to auto JIT functions, but I think that's way out of scope for this issue.

One more thought on all of this: we might want to invent our own dialect that maps 1:1 with our AST. The upside there being that AST lowering within rust is dead simple, and the work to lower from FishDialect down to LLVM can be done just like any other MLIR dialect.

---

The linked issue above mentioned that there's quite a bit of work to get this going. I agree, but notably I'm interested in doing that work. Before sinking too much time into this, I'm curious what the community's thoughts are.

**Comment by faho** [comment]
I'm sorry, but I am going to repeat what I said in #4316:

There are many better things to spend our time on that will have a more immmediate impact on performance. Chief among them not constantly encoding to UTF-32 and back, and allowing builtins to run in parallel.

Then there's also the fact that long-running shellscripts will often spend most of their time running external commands that no JIT in the shell will ever affect - [Amdahl's Law](https://en.wikipedia.org/wiki/Amdahl%27s_law).

In addition there's the additional code complexity and dependencies we would be taking on.

Overall, I do not believe this is a good task to spend effort on and I don't believe it is all that promising. I would recommend you would try to find something else to work on. Of course you can always prove me wrong, but until then I will be closing this.
